lib/Service/DownloadService.php

```diff
@@ -2,25 +2,117 @@
-use GuzzleHttp\Client;
-use GuzzleHttp\Promise\Utils;
+use Exception;
+use InvalidArgumentException;
+use JetBrains\PhpStorm\NoReturn;
+use OCA\OpenRegister\Db\RegisterMapper;
+use OCA\OpenRegister\Db\SchemaMapper;
 use OCP\IURLGenerator;
-use Symfony\Component\Uid\Uuid;
 
 class DownloadService
 {
-	public function __construct() {}
+	public function __construct(
+		private IURLGenerator $urlGenerator,
+		private SchemaMapper $schemaMapper,
+		private RegisterMapper $registerMapper
+	) {}
 
-	public function download(string $type)
+	/**
+	 * Download a DB entity as a file. Depending on given Accept-header the file type might differ.
+	 *
+	 * @param string $objectType The type of object to download.
+	 * @param string|int $id The id of the object to download.
+	 * @param string $accept The Accept-header from the download request.
+	 *
+	 * @return array The response data for the download request.
+	 * @throws Exception
+	 */
+	public function download(string $objectType, string|int $id, string $accept)
 	{
-		switch ($type) {
-			case 'json':
-				// @todo this is placeholder code
-				break;
-			default:
-				// @todo some logging
-				return null;
+		// Get the appropriate mapper for the object type
+		$mapper = $this->getMapper($objectType);
+
+		try {
+			$object = $mapper->find($id);
+		} catch (Exception $exception) {
+			return ['error' => "Could not find $objectType with id $id.", 'statusCode' => 404];
+		}
+
+		$objectArray = $object->jsonSerialize();
+		$filename = $objectArray['title'].ucfirst($objectType).'-v'.$objectArray['version'];
+
+		if (str_contains(haystack: $accept, needle: 'application/json') === true) {
+			$url = $this->urlGenerator->getAbsoluteURL($this->urlGenerator->linkToRoute('openregister.'.ucfirst($objectType).'s.show', ['id' => $object->getId()]));
+
+			// @todo: json+ld? str_contains($accept,'json+ld') === true, else:
+			// Json...
+			$objArray['title'] = $objectArray['title'];
+			$objArray['$id'] = $url;
+//			$objArray['$schema'] = 'https://json-schema.org/draft/2020-12/schema';
+			$objArray['$schema'] = 'https://docs.commongateway.nl/schemas/'.ucfirst($objectType).'.schema.json';
+			$objArray['version'] = $objectArray['version'];
+			$objArray['type'] = $objectType;
+			unset($objectArray['title'], $objectArray['version'], $objectArray['id'], $objectArray['uuid']);
+			$objArray = array_merge($objArray, $objectArray);
+
+			// Convert the object data to JSON
+			$jsonData = json_encode($objArray, JSON_PRETTY_PRINT);
+
+			$this->downloadJson($jsonData, $filename);
 		}
+
+		return ['error' => "The Accept type $accept is not supported.", 'statusCode' => 400];
+	}
+
+	/**
+	 * Generate a downloadable json file response.
+	 *
+	 * @param string $jsonData The json data to create a json file with.
+	 * @param string $filename The filename, .json will be added after this filename in this function.
+	 *
+	 * @return void
+	 */
+	#[NoReturn] private function downloadJson(string $jsonData, string $filename): void
```

Comment:
#[NoReturn] is nieuw voor mij, kan je deze toelichten?

Reply:
Automatisch aangevuld door PHPStorm en staat wel mooi. En deze dingen:
![image](https://github.com/user-attachments/assets/ad83db86-4926-486a-8405-411ef07a13a4)

Reply:
En er is daadwerkelijk een functioneel verschil tussen dit en : void als return type opgeven bij de functie
![image](https://github.com/user-attachments/assets/aa9d131f-0871-4f8c-881f-f190655237c3)